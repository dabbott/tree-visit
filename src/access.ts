import { convertChildrenToEntries, getEntriesChild } from './getChild'
import { IndexPath, KeyPath } from './indexPath'
import { BaseEntriesOptions, BaseOptions } from './options'

type AccessChildrenOptions<T> = BaseOptions<T> & {
  indexPath: IndexPath
}

type AccessEntriesOptions<T> = BaseEntriesOptions<T> & {
  keyPath: KeyPath
}

export type AccessOptions<T> =
  | AccessChildrenOptions<T>
  | AccessEntriesOptions<T>

/**
 * Returns a node by its `IndexPath`.
 *
 * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
 */
export function access<T>(node: T, options: AccessChildrenOptions<T>): T
export function access<T>(node: T, options: AccessEntriesOptions<T>): T
export function access<T>(node: T, options: AccessOptions<T>): T {
  const opts =
    'getChildren' in options
      ? { ...convertChildrenToEntries<T>(options), keyPath: options.indexPath }
      : options

  return accessInternal(node, opts)
}

export function accessInternal<T>(
  node: T,
  options: AccessEntriesOptions<T>
): T {
  let path = options.keyPath.slice()

  while (path.length > 0) {
    let key = path.shift()!

    node = getEntriesChild(node, {
      ...options,
      keyPath: path,
      key,
    })
  }

  return node
}

/**
 * Returns an array of each node in an `IndexPath`.
 *
 * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
 */
export function accessPath<T>(node: T, options: AccessChildrenOptions<T>): T[]
export function accessPath<T>(node: T, options: AccessEntriesOptions<T>): T[]
export function accessPath<T>(node: T, options: AccessOptions<T>): T[] {
  const opts =
    'getChildren' in options
      ? { ...convertChildrenToEntries<T>(options), keyPath: options.indexPath }
      : options

  return accessPathInternal(node, opts)
}

export function accessPathInternal<T>(
  node: T,
  options: AccessEntriesOptions<T>
): T[] {
  let result: T[] = [node]
  let path = options.keyPath.slice()

  while (path.length > 0) {
    let key = path.shift()!

    node = getEntriesChild(node, {
      ...options,
      keyPath: path,
      key,
    })

    result.push(node)
  }

  return result
}
