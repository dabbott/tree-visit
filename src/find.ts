import { IndexPath } from './indexPath'
import { BaseChildrenOptions } from './options'
import { STOP, visit } from './visit'

export type FindOptions<T> = BaseChildrenOptions<T> & {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, indexPath: IndexPath) => boolean
}

export type FindOptionsTyped<T, S extends T> = BaseChildrenOptions<T> & {
  /**
   * Return `true` to include this node in the results.
   */
  predicate: (node: T, indexPath: IndexPath) => node is S
}

/**
 * Find a node matching a predicate function.
 */
export function find<T>(node: T, options: FindOptions<T>): T | undefined
export function find<T, S extends T>(
  node: T,
  options: FindOptionsTyped<T, S>
): S | undefined
export function find<T>(node: T, options: FindOptions<T>): T | undefined {
  let found: T | undefined

  visit(node, {
    ...options,
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        found = child
        return STOP
      }
    },
  })

  return found
}

/**
 * Find all nodes matching a predicate function.
 */
export function findAll<T>(node: T, options: FindOptions<T>): T[]
export function findAll<T, S extends T>(
  node: T,
  options: FindOptionsTyped<T, S>
): S[]
export function findAll<T>(node: T, options: FindOptions<T>): T[] {
  let found: T[] = []

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        found.push(child)
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

/**
 * Find the `IndexPath` of a node matching a predicate function.
 */
export function findIndexPath<T>(
  node: T,
  options: FindOptions<T>
): IndexPath | undefined {
  let found: IndexPath | undefined

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        // Copy the indexPath, since indexPath may be mutated
        found = [...indexPath]
        return STOP
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

/**
 * Find the `IndexPath` of all nodes matching a predicate function.
 */
export function findAllIndexPaths<T>(
  node: T,
  options: FindOptions<T>
): IndexPath[] {
  let found: IndexPath[] = []

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        // Copy the indexPath, since indexPath may be mutated
        found.push([...indexPath])
      }
    },
    getChildren: options.getChildren,
  })

  return found
}
