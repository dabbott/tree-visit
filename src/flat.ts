import { BaseOptions } from './options'
import { visit } from './visit'

/**
 * Returns an array containing the root node and all of its descendants.
 *
 * This is analogous to `Array.prototype.flat` for flattening arrays.
 */
export function flat<T>(node: T, options: BaseOptions<T>): T[] {
  let nodes: T[] = []

  visit(node, {
    onEnter: (child) => {
      nodes.push(child)
    },
    getChildren: options.getChildren,
  })

  return nodes
}
