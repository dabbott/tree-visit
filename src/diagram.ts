import { boxDiagram } from './diagram/boxDiagram'
import { directoryDiagram } from './diagram/directoryDiagram'
import { IndexPath } from './indexPath'
import { BaseChildrenOptions } from './options'

export type DiagramType = 'directory' | 'box'

export type DiagramOptions<T> = BaseChildrenOptions<T> & {
  getLabel: (node: T, indexPath: IndexPath) => string
  type?: DiagramType
  flattenSingleChildNodes?: boolean
}

/**
 * Generate a diagram of the tree, as a string.
 */
export function diagram<T>(node: T, options: DiagramOptions<T>): string {
  if (options.type === 'box') {
    return boxDiagram(node, options)
  }

  return directoryDiagram(node, options)
}
