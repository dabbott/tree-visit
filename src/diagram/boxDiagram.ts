import { DiagramEntriesOptions } from '../diagram'

enum BoxDrawing {
  TopLeft = '┌',
  TopRight = '┐',
  BottomLeft = '└',
  BottomRight = '┘',
  Horizontal = '─',
  Vertical = '│',
  BottomConnectorDown = '┬',
  BottomConnectorUp = '┴',
  TopConnectorUp = '┴',
}

type Box = {
  width: number
  height: number
  contents: string[]
}

function wrapLabelInBox(label: string) {
  const lines = label.split('\n')
  const length = Math.max(...lines.map((line) => line.length))
  const horizontalMargin = 1
  const width = length + horizontalMargin * 2 + 2
  const height = 2 + lines.length
  const diagram = [
    [
      BoxDrawing.TopLeft,
      BoxDrawing.Horizontal.repeat(length + horizontalMargin * 2),
      BoxDrawing.TopRight,
    ],
    ...lines.map((line) => [
      BoxDrawing.Vertical,
      ' '.repeat(horizontalMargin),
      line + (line.length < length ? ' '.repeat(length - line.length) : ''),
      ' '.repeat(horizontalMargin),
      BoxDrawing.Vertical,
    ]),
    [
      BoxDrawing.BottomLeft,
      BoxDrawing.Horizontal.repeat(length + horizontalMargin * 2),
      BoxDrawing.BottomRight,
    ],
  ]

  return {
    width,
    height,
    contents: diagram.map((parts) => parts.join('')),
  }
}

function mergeBoxesHorizontal(boxes: Box[]): Box {
  const horizontalMargin = 1

  if (boxes.length === 0) {
    throw new Error("Can't merge empty array of boxes")
  }

  return boxes.slice(1).reduce((result, box) => {
    const height = Math.max(result.height, box.height)
    const width = result.width + horizontalMargin + box.width
    const contents: string[] = []

    for (let i = 0; i < height; i++) {
      contents.push(
        (result.contents[i] || ' '.repeat(result.width)) +
          ' '.repeat(horizontalMargin) +
          (box.contents[i] || ' '.repeat(box.width))
      )
    }

    return { height, width, contents: centerBlock(contents, width) }
  }, boxes[0])
}

function mergeBoxesVertical(boxes: Box[]): Box {
  const verticalMargin = 1

  if (boxes.length === 0) {
    throw new Error("Can't merge empty array of boxes")
  }

  const width = Math.max(...boxes.map((box) => box.width))

  return boxes.slice(1).reduce((result, box) => {
    const height = result.height + verticalMargin + box.height
    const contents: string[] = []

    for (let i = 0; i < height; i++) {
      if (i < result.height) {
        contents.push(result.contents[i])
      } else if (i === result.height) {
        contents.push(' '.repeat(width))
      } else {
        contents.push(box.contents[i - result.height - 1])
      }
    }

    return { height, width, contents: centerBlock(contents, width) }
  }, boxes[0])
}

function nodeDiagram<T, PK extends PropertyKey>(
  node: T,
  keyPath: PK[],
  options: DiagramEntriesOptions<T, PK>
): Box {
  const label = options.getLabel(node, keyPath)
  const box = wrapLabelInBox(label)

  const entries = options.getEntries(node, keyPath)

  if (entries.length === 0) return box

  const childBoxes = entries.map(([key, child]) => {
    const childBox = nodeDiagram(child, [...keyPath, key], options)

    // Draw top connector for each child
    childBox.contents[0] = insertSubstring(
      childBox.contents[0],
      centerIndex(childBox.contents[0].length),
      BoxDrawing.TopConnectorUp
    )

    return childBox
  })

  const result = mergeBoxesVertical([box, mergeBoxesHorizontal(childBoxes)])

  // Draw connectors
  const contents = result.contents
  const mid = centerIndex(result.width)

  contents[box.height - 1] = insertSubstring(
    contents[box.height - 1],
    mid,
    BoxDrawing.BottomConnectorDown
  )

  let min = contents[box.height + 1].indexOf(BoxDrawing.TopConnectorUp)
  let max = contents[box.height + 1].lastIndexOf(BoxDrawing.TopConnectorUp)

  for (let i = min; i <= max; i++) {
    let character: string

    if (i === mid) {
      character =
        childBoxes.length > 1
          ? BoxDrawing.BottomConnectorUp
          : BoxDrawing.Vertical
    } else if (i === min) {
      character = BoxDrawing.TopLeft
    } else if (i === max) {
      character = BoxDrawing.TopRight
    } else {
      character = BoxDrawing.Horizontal
    }

    result.contents[box.height] = insertSubstring(
      result.contents[box.height],
      i,
      character
    )
  }

  return result
}

export function boxDiagram<T, PK extends PropertyKey>(
  node: T,
  options: DiagramEntriesOptions<T, PK>
): string {
  return nodeDiagram(node, [], options).contents.join('\n')
}

function centerIndex(width: number): number {
  return Math.floor(width / 2)
}

function centerLine(line: string, width: number): string {
  const remainder = width - line.length

  if (remainder <= 0) return line

  const prefixLength = centerIndex(remainder)
  const suffixLength = centerIndex(remainder)
  const result = ' '.repeat(prefixLength) + line + ' '.repeat(suffixLength)

  return prefixLength + suffixLength + result.length < width
    ? result + ' '
    : result
}

function centerBlock(lines: string[], width: number): string[] {
  return lines.map((line) => centerLine(line, width))
}

function insertSubstring(string: string, index: number, substring: string) {
  if (index > string.length - 1) return string

  return string.substring(0, index) + substring + string.substring(index + 1)
}
