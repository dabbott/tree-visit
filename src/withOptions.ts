import { BaseOptions } from './options'
import { visit, VisitOptions } from './visit'
import { IndexPath } from './indexPath'
import { access, accessPath } from './access'
import { flat } from './flat'
import {
  find,
  findAll,
  findAllIndexPaths,
  findIndexPath,
  FindOptions,
  FindOptionsTyped,
} from './find'
import { DiagramOptions, diagram } from './diagram'

// Omit any keys from the BaseOptions, since they're always applied
// automatically as part of `withOptions`
type WithoutBase<O> = Omit<O, keyof BaseOptions<unknown>>

export type WithOptions<T> = {
  /**
   * Returns a node by its `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  access(node: T, indexPath: IndexPath): T

  /**
   * Returns an array of each node in an `IndexPath`.
   *
   * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
   */
  accessPath(node: T, indexPath: IndexPath): T[]

  /**
   * Generate a diagram of the tree, as a string.
   */
  diagram(node: T, getLabel: DiagramOptions<T>['getLabel']): string
  diagram(node: T, options: WithoutBase<DiagramOptions<T>>): string

  /**
   * Find a node matching a predicate function.
   */
  find(node: T, predicate: FindOptions<T>['predicate']): T | undefined
  find(node: T, options: WithoutBase<FindOptions<T>>): T | undefined
  find<S extends T>(
    node: T,
    predicate: FindOptionsTyped<T, S>['predicate']
  ): S | undefined
  find<S extends T>(
    node: T,
    options: WithoutBase<FindOptionsTyped<T, S>>
  ): S | undefined

  /**
   * Find all nodes matching a predicate function.
   */
  findAll(node: T, predicate: FindOptions<T>['predicate']): T[]
  findAll(node: T, options: WithoutBase<FindOptions<T>>): T[]
  findAll<S extends T>(
    node: T,
    predicate: FindOptionsTyped<T, S>['predicate']
  ): S[]
  findAll<S extends T>(
    node: T,
    options: WithoutBase<FindOptionsTyped<T, S>>
  ): S[]

  /**
   * Find the `IndexPath` of a node matching a predicate function.
   */
  findIndexPath(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath | undefined
  findIndexPath(
    node: T,
    options: WithoutBase<FindOptions<T>>
  ): IndexPath | undefined

  /**
   * Find the `IndexPath` of all nodes matching a predicate function.
   */
  findAllIndexPaths(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath[]
  findAllIndexPaths(node: T, options: WithoutBase<FindOptions<T>>): IndexPath[]

  /**
   * Returns an array containing the root node and all of its descendants.
   *
   * This is analogous to `Array.prototype.flat` for flattening arrays.
   */
  flat(node: T): T[]

  /**
   * Visit each node using preorder DFS traversal.
   */
  visit(node: T, onEnter: NonNullable<VisitOptions<T>>['onEnter']): void

  /**
   * Visit each node using DFS traversal.
   */
  visit(node: T, options: WithoutBase<VisitOptions<T>>): void
}

/**
 * Return every tree utility function with options partially applied.
 *
 * @param baseOptions
 */
export function withOptions<T>(baseOptions: BaseOptions<T>): WithOptions<T> {
  return {
    access: (node, indexPath) => access(node, indexPath, baseOptions),
    accessPath: (node, indexPath) => accessPath(node, indexPath, baseOptions),
    diagram: (
      node: T,
      getLabelOrOptions:
        | DiagramOptions<T>['getLabel']
        | WithoutBase<DiagramOptions<T>>
    ) =>
      typeof getLabelOrOptions === 'function'
        ? diagram(node, { ...baseOptions, getLabel: getLabelOrOptions })
        : diagram(node, { ...baseOptions, ...getLabelOrOptions }),
    find: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? find(node, { ...baseOptions, predicate: predicateOrOptions })
        : find(node, { ...baseOptions, ...predicateOrOptions }),
    findAll: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findAll(node, { ...baseOptions, predicate: predicateOrOptions })
        : findAll(node, { ...baseOptions, ...predicateOrOptions }),
    findIndexPath: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findIndexPath(node, { ...baseOptions, predicate: predicateOrOptions })
        : findIndexPath(node, { ...baseOptions, ...predicateOrOptions }),
    flat: (node: T) => flat(node, baseOptions),
    findAllIndexPaths: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | WithoutBase<FindOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findAllIndexPaths(node, {
            ...baseOptions,
            predicate: predicateOrOptions,
          })
        : findAllIndexPaths(node, { ...baseOptions, ...predicateOrOptions }),
    visit: (
      node: T,
      onEnterOrOptions:
        | NonNullable<VisitOptions<T>>['onEnter']
        | WithoutBase<VisitOptions<T>>
    ) =>
      typeof onEnterOrOptions === 'function'
        ? visit(node, { ...baseOptions, onEnter: onEnterOrOptions })
        : visit(node, { ...baseOptions, ...onEnterOrOptions }),
  }
}
