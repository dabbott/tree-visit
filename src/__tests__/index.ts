import { visit, IndexPath } from '../index'
import { access, accessPath } from '../access'
import { find, findIndexPath, findAll, findAllIndexPaths } from '../find'
import { withOptions } from '../withOptions'
import { diagram } from '../diagram'
import { flat } from '../flat'

type Node = {
  name: string
  children?: Node[]
  indexPath: IndexPath
}

function getChildren(node: Node): Node[] {
  return node.children ?? []
}

const example: Node = {
  name: 'a',
  children: [
    {
      name: 'b',
      indexPath: [0],
      children: [
        { name: 'b1', indexPath: [0, 0] },
        { name: 'b2', indexPath: [0, 1] },
      ],
    },
    {
      name: 'c',
      indexPath: [1],
      children: [
        { name: 'c1', indexPath: [1, 0] },
        { name: 'c2', indexPath: [1, 1] },
      ],
    },
  ],
  indexPath: [],
}

type TypeA = Node & {
  type: 'a'
}
type TypeB = Node & {
  type: 'b'
}

// We use this union of types to check if we successfully narrow the
// type to a specific one of them using type predicates. TypeScript will
// throw an error when compiling if we try to check e.g. type === 'a' on
// an object of TypeB
type TypedExample = TypeA | TypeB

function getTypedChildren(typedNode: TypedExample): TypedExample[] {
  return getChildren(typedNode).map((item) => ({ type: 'b', ...item }))
}

const typedExample: TypedExample = {
  type: 'a',
  ...example,
}

function createNestedNode(depth: number): Node {
  let count = 0

  let root: Node = { name: `${count++}`, indexPath: [], children: [] }
  let current = root

  for (let i = 0; i < depth - 1; i++) {
    let child: Node = {
      name: `${count++}`,
      indexPath: [...current.indexPath, 0],
      children: [],
    }

    current.children?.push(child)

    current = child
  }

  return root
}

it('traverses normally', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []
  let getChildrenCount = 0

  let countGetChildren = (node: Node) => {
    getChildrenCount++
    return getChildren(node)
  }

  visit(example, {
    onEnter: (child, indexPath) => {
      expect(indexPath).toEqual(child.indexPath)
      enterNames.push(child.name)
    },
    onLeave: (child, indexPath) => {
      expect(indexPath).toEqual(child.indexPath)
      leaveNames.push(child.name)
    },
    getChildren: countGetChildren,
  })

  expect(getChildrenCount).toEqual(7)
  expect(enterNames).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])
  expect(leaveNames).toEqual(['b1', 'b2', 'b', 'c1', 'c2', 'c', 'a'])
})

it('traverses deeply nested nodes', () => {
  let enterNames: string[] = []

  const nestedNode = createNestedNode(6000)

  visit(nestedNode, {
    onEnter: (child) => {
      enterNames.push(child.name)
    },
    getChildren,
  })

  expect(enterNames.length).toEqual(6000)
})

it('skips in onEnter', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child) => {
      enterNames.push(child.name)
      if (child.name === 'b') return 'skip'
    },
    onLeave: (child) => {
      leaveNames.push(child.name)
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b', 'c', 'c1', 'c2'])
  expect(leaveNames).toEqual(['c1', 'c2', 'c', 'a'])
})

it('stops in onEnter', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child) => {
      enterNames.push(child.name)
      if (child.name === 'b') return 'stop'
    },
    onLeave: (child) => {
      leaveNames.push(child.name)
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b'])
  expect(leaveNames).toEqual([])
})

it('stops in onLeave', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child) => {
      enterNames.push(child.name)
    },
    onLeave: (child) => {
      leaveNames.push(child.name)
      if (child.name === 'b') return 'stop'
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b', 'b1', 'b2'])
  expect(leaveNames).toEqual(['b1', 'b2', 'b'])
})

describe('access', () => {
  it('accesses node at index path', () => {
    visit(example, {
      onEnter: (child, indexPath) => {
        const found = access(example, indexPath, { getChildren })
        expect(found.name).toEqual(child.name)
      },
      getChildren,
    })
  })

  it('accesses node path', () => {
    expect(
      accessPath(example, [0, 1], { getChildren }).map((node) => node.name)
    ).toEqual(['a', 'b', 'b2'])
  })
})

describe('find', () => {
  it('finds a node', () => {
    const node = find(example, {
      getChildren,
      predicate: (node) => node.name === 'b1',
    })

    expect(node?.name).toEqual('b1')
  })

  it('finds a typed node', () => {
    const node = find<TypedExample, TypeA>(typedExample, {
      getChildren: getTypedChildren,
      predicate: (node): node is TypeA => node.type === 'a',
    })

    // This will throw a compiler error if not working correctly
    expect(node && node.type === 'a').toEqual(true)
  })

  it('finds all nodes', () => {
    const nodes = findAll(example, {
      getChildren,
      predicate: (node) => node.name.startsWith('b'),
    })

    expect(nodes.map((node) => node.name)).toEqual(['b', 'b1', 'b2'])
  })

  it('finds all typed nodes', () => {
    const nodes = findAll<TypedExample, TypeA>(typedExample, {
      getChildren: getTypedChildren,
      predicate: (node): node is TypeA => node.type === 'a',
    })

    // This will throw a compiler error if not working correctly
    expect(nodes[0].type === 'a').toEqual(true)
  })

  it('finds a node index path', () => {
    const indexPath = findIndexPath(example, {
      getChildren,
      predicate: (node, indexPath) => indexPath.join() === [0, 1].join(),
    })

    expect(indexPath).toEqual([0, 1])
  })

  it('finds all node index paths', () => {
    const indexPaths = findAllIndexPaths(example, {
      getChildren,
      predicate: (node) => node.name.startsWith('b'),
    })

    expect(indexPaths).toEqual([[0], [0, 0], [0, 1]])

    const nodes = indexPaths.map((indexPath) =>
      access(example, indexPath, { getChildren })
    )

    expect(nodes.map((node) => node.name)).toEqual(['b', 'b1', 'b2'])
  })
})

describe('flat', () => {
  it('flattens a tree', () => {
    const nodes = flat(example, {
      getChildren,
    })

    expect(nodes.map((node) => node.name)).toEqual([
      'a',
      'b',
      'b1',
      'b2',
      'c',
      'c1',
      'c2',
    ])
  })
})

describe('diagram', () => {
  const getLabel = (node: Node): string => node.name
  const getMultilineLabel = (node: Node): string =>
    `name: ${node.name}\npath: ${node.indexPath.join('.')}`

  it('generates folder diagram', () => {
    expect(diagram(example, { getChildren, getLabel })).toMatchSnapshot()

    expect(
      diagram(example, { getChildren, getLabel, flattenSingleChildNodes: true })
    ).toMatchSnapshot()
  })

  it('generates folder diagram with single child case', () => {
    const singleChild: Node = {
      name: 'a',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(diagram(singleChild, { getChildren, getLabel })).toMatchSnapshot()

    expect(
      diagram(singleChild, {
        getChildren,
        getLabel,
        flattenSingleChildNodes: true,
      })
    ).toMatchSnapshot()
  })

  it('generates folder diagram with non-root single child case', () => {
    const singleChild: Node = {
      name: 'a',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
        {
          name: 'c',
          indexPath: [1],
          children: [{ name: 'c1', indexPath: [1, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(diagram(singleChild, { getChildren, getLabel })).toMatchSnapshot()

    expect(
      diagram(singleChild, {
        getChildren,
        getLabel,
        flattenSingleChildNodes: true,
      })
    ).toMatchSnapshot()
  })

  it('generates folder diagram with hidden root', () => {
    const singleChild: Node = {
      name: '',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(diagram(singleChild, { getChildren, getLabel })).toMatchSnapshot()

    expect(
      diagram(singleChild, {
        getChildren,
        getLabel,
        flattenSingleChildNodes: true,
      })
    ).toMatchSnapshot()
  })

  it('generates folder diagram with multiline label', () => {
    expect(
      diagram(example, { getChildren, getLabel: getMultilineLabel })
    ).toMatchSnapshot()

    expect(
      diagram(example, {
        getChildren,
        getLabel: getMultilineLabel,
        flattenSingleChildNodes: true,
      })
    ).toMatchSnapshot()
  })

  it('generates folder diagram with multiline label and single child case', () => {
    const singleChild: Node = {
      name: 'a',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(
      diagram(singleChild, { getChildren, getLabel: getMultilineLabel })
    ).toMatchSnapshot()

    expect(
      diagram(singleChild, { getChildren, getLabel: getMultilineLabel })
    ).toMatchSnapshot()
  })

  it('generates box diagram', () => {
    expect(
      diagram(example, { type: 'box', getChildren, getLabel })
    ).toMatchSnapshot()
  })

  it('generates multiline box diagram', () => {
    expect(
      diagram(example, {
        type: 'box',
        getChildren,
        getLabel: (node: Node): string =>
          node.name === 'b'
            ? `name: ${node.name}\npath: ${node.indexPath.join('.')}`
            : `name: ${node.name}`,
      })
    ).toMatchSnapshot()
  })

  it('generates uneven box diagram', () => {
    const node: Node = {
      name: 'a',
      children: [
        {
          name: 'hello',
          indexPath: [0],
        },
        {
          name: 'c',
          indexPath: [1],
        },
      ],
      indexPath: [],
    }

    expect(
      diagram(node, { type: 'box', getChildren, getLabel })
    ).toMatchSnapshot()
  })

  it('generates single child box', () => {
    const node: Node = {
      name: 'a',
      children: [
        {
          name: 'okay',
          indexPath: [0],
        },
      ],
      indexPath: [],
    }

    expect(
      diagram(node, { type: 'box', getChildren, getLabel })
    ).toMatchSnapshot()
  })
})

describe('withOptions', () => {
  it('binds options', () => {
    const { find, access, visit } = withOptions({ getChildren })

    let enterNames: string[] = []

    visit(example, {
      onEnter: (child) => {
        enterNames.push(child.name)
      },
    })

    expect(enterNames).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])

    expect(
      find(example, {
        predicate: (node) => node.name === 'b1',
      })?.name
    ).toEqual('b1')

    expect(access(example, [0, 0]).name).toEqual('b1')
  })

  it('supports typed finding', () => {
    const { find, findAll } = withOptions({ getChildren: getTypedChildren })

    expect(
      find<TypeA>(typedExample, {
        predicate: (node): node is TypeA => node.type === 'a',
      })?.type === 'a'
    ).toEqual(true)

    expect(
      findAll<TypeA>(typedExample, {
        predicate: (node): node is TypeA => node.type === 'a',
      })[0]?.type === 'a'
    ).toEqual(true)
  })

  it('supports overloaded calls', () => {
    const { find, findAllIndexPaths, visit, diagram, flat } = withOptions({
      getChildren,
    })

    let enterNames: string[] = []

    visit(example, (child) => {
      enterNames.push(child.name)
    })

    expect(enterNames).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])

    expect(flat(example).map((node) => node.name)).toEqual([
      'a',
      'b',
      'b1',
      'b2',
      'c',
      'c1',
      'c2',
    ])

    expect(find(example, (node) => node.name === 'b1')?.name).toEqual('b1')

    expect(
      findAllIndexPaths(example, (node) => node.name.startsWith('b'))
    ).toEqual([[0], [0, 0], [0, 1]])

    expect(diagram(example, (node) => node.name)).toMatchSnapshot()

    expect(
      diagram(example, {
        getLabel: (node) => node.name,
      })
    ).toMatchSnapshot()
  })
})
