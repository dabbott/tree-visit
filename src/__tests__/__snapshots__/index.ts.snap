// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diagram generates box diagram 1`] = `
"           ┌───┐           
           │ a │           
           └─┬─┘           
      ┌──────┴──────┐      
    ┌─┴─┐         ┌─┴─┐    
    │ b │         │ c │    
    └─┬─┘         └─┬─┘    
   ┌──┴───┐      ┌──┴───┐  
┌──┴─┐ ┌──┴─┐ ┌──┴─┐ ┌──┴─┐
│ b1 │ │ b2 │ │ c1 │ │ c2 │
└────┘ └────┘ └────┘ └────┘"
`;

exports[`diagram generates folder diagram 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`diagram generates folder diagram 2`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`diagram generates folder diagram with hidden root 1`] = `
"
└── b
    └── b1"
`;

exports[`diagram generates folder diagram with hidden root 2`] = `"/ b / b1"`;

exports[`diagram generates folder diagram with multiline label 1`] = `
"name: a
path: 
├── name: b
│   path: 0
│   ├── name: b1
│   │   path: 0.0
│   └── name: b2
│       path: 0.1
└── name: c
    path: 1
    ├── name: c1
    │   path: 1.0
    └── name: c2
        path: 1.1"
`;

exports[`diagram generates folder diagram with multiline label 2`] = `
"name: a
path: 
├── name: b
│   path: 0
│   ├── name: b1
│   │   path: 0.0
│   └── name: b2
│       path: 0.1
└── name: c
    path: 1
    ├── name: c1
    │   path: 1.0
    └── name: c2
        path: 1.1"
`;

exports[`diagram generates folder diagram with multiline label and single child case 1`] = `
"name: a
path: 
└── name: b
    path: 0
    └── name: b1
        path: 0.0"
`;

exports[`diagram generates folder diagram with multiline label and single child case 2`] = `
"name: a
path: 
└── name: b
    path: 0
    └── name: b1
        path: 0.0"
`;

exports[`diagram generates folder diagram with non-root single child case 1`] = `
"a
├── b
│   └── b1
└── c
    └── c1"
`;

exports[`diagram generates folder diagram with non-root single child case 2`] = `
"a
├── b / b1
└── c / c1"
`;

exports[`diagram generates folder diagram with single child case 1`] = `
"a
└── b
    └── b1"
`;

exports[`diagram generates folder diagram with single child case 2`] = `"a / b / b1"`;

exports[`diagram generates multiline box diagram 1`] = `
"                    ┌─────────┐                    
                    │ name: a │                    
                    └────┬────┘                    
            ┌────────────┴────────────┐            
       ┌────┴────┐               ┌────┴────┐       
       │ name: b │               │ name: c │       
       │ path: 0 │               └────┬────┘       
       └────┬────┘              ┌─────┴──────┐     
      ┌─────┴──────┐      ┌─────┴────┐ ┌─────┴────┐
┌─────┴────┐ ┌─────┴────┐ │ name: c1 │ │ name: c2 │
│ name: b1 │ │ name: b2 │ └──────────┘ └──────────┘
└──────────┘ └──────────┘                          "
`;

exports[`diagram generates single child box 1`] = `
" ┌───┐ 
 │ a │ 
 └──┬┘ 
    │   
┌───┴──┐
│ okay │
└──────┘"
`;

exports[`diagram generates uneven box diagram 1`] = `
"     ┌───┐     
     │ a │     
     └─┬─┘     
    ┌──┴────┐  
┌───┴───┐ ┌─┴─┐
│ hello │ │ c │
└───────┘ └───┘"
`;

exports[`withOptions supports overloaded calls 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`withOptions supports overloaded calls 2`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;
